Geometry: New sample

Jade-JadeH opened the pull request:
**Sample**

geometry sample corresponding to the unity version

- [x] Match function and variable name with Unity
- [x] Match UI with Unity
- [x] Modify input manager to support on click end and dragging interaction; disable pawn during pressing shift to support natural interaction
- [x] Remove input manager variable initialization from .h to .cpp for code safety, change variable name to UpperCamelCase, `mousePress` to `MousePress` 
- [x] One thing that is left is the notification which still shows the measurement sample notification, Andy will send me the png to replace. 

**Checklist**

<!--- Delete any that don't apply -->

- [x] PR title follows convention - `keyword: Short description of change` <!--- Example:  ansible: Update build scripts to handle new engine versions -->
- [x] PR targets the correct branch <!-- Changes going into `main` work with the current public release of the plugin-->
- [x] New Blueprints follow Unreal Naming Convention <!-- bp_BlueprintFileName, wbp_WidgetFileName, e_EnumFileName, animbp_AnimationBlueprintFileName, etc -->
- [x] Self-review of changes
- [x] There are no warnings related to changes
- [ ] A build was made and tested on all relevant platforms
- [x] No unrelated changes have been made to any other code or project files
- [x] No unnecessary includes or namespaces added
- [x] Code follows plugin coding style
- [x] New code and changed code has proper formatting <!-- Run a formatter if unsure -->
- [x] No unintentional formatting changes
- [x] Commits have descriptive titles

**Blueprints**
Change this bp: 
![before](https://github.com/user-attachments/assets/0a3909e3-d98a-4a1a-b91a-2a7b05c03ac6)

To this one: 
![after](https://github.com/user-attachments/assets/b6c62744-2598-4bc5-a4fe-0644a5ca5cb0)

**ArcGIS Maps SDK Version**

2.0+5.3

mmgaw113 commented on sample_project/Source/sample_project/InputManager.h:
```diff
@@ -20,20 +24,39 @@ class SAMPLE_PROJECT_API AInputManager : public AActor
+	UPROPERTY(BlueprintAssignable)
+	FOnInputEnd OnInputEnd;
+	
+	UFUNCTION()
+	void OnShiftPressed();
+
+	UFUNCTION()
+	void OnShiftReleased();
+
+	UFUNCTION()
+	void TriggerInputStart(); 
+
+	UFUNCTION()
+	void TriggerInputEnd();
+
```
Is there a reason these are public and UFunctions? They shouldn't need to be because of the events.

Jade-JadeH replied:
I think usually these input functions are ufunctions.

mmgaw113 replied:
Unless they are being used in blueprints or in other files, they don't need to be ufunctions and public. They can be private normal functions.

Jade-JadeH replied:
was referring to this sentence: "UFunctions are able to bind to [Delegates](https://dev.epicgames.com/documentation/en-us/unreal-engine/delegates?application_version=4.27) in the default properties of a Class, enabling them to perform such tasks as associating actions with user inputs."

mmgaw113 commented on sample_project/Source/sample_project/InputManager.h:
```diff
@@ -20,20 +24,39 @@ class SAMPLE_PROJECT_API AInputManager : public AActor
+	UFUNCTION()
+	void OnShiftReleased();
+
+	UFUNCTION()
+	void TriggerInputStart(); 
+
+	UFUNCTION()
+	void TriggerInputEnd();
+
 protected:
 	// Called when the game starts or when spawned
 	virtual void BeginPlay() override;
 	virtual void EndPlay(const EEndPlayReason::Type EndPlayReason) override;
 	virtual void SetupPlayerInputComponent(UInputComponent* PlayerInputComponent);
 
 private:
-	void TriggerInput();
-	
+
 	UPROPERTY(BlueprintReadOnly, VisibleAnywhere, meta = (AllowPrivateAccess))
-	UInputMappingContext* MappingContext = LoadObject<UInputMappingContext>(nullptr, TEXT("/Script/EnhancedInput.InputMappingContext'/Game/SampleViewer/SharedResources/Input/IAC_SamplesInput.IAC_SamplesInput'"));
+	UInputMappingContext* MappingContext;
+
+	UPROPERTY(BlueprintReadOnly, VisibleAnywhere, meta = (AllowPrivateAccess))
+	UInputAction* MousePress;
+
 	UPROPERTY(BlueprintReadOnly, VisibleAnywhere, meta = (AllowPrivateAccess))
-	UInputAction* mousePress = LoadObject<UInputAction>(nullptr, TEXT("/Script/EnhancedInput.InputAction'/Game/SampleViewer/SharedResources/Input/IA_LeftMouseClick.IA_LeftMouseClick'"));
+	UInputAction* ShiftModifier;
```
See above regarding keeping load object in header file.

mmgaw113 commented on sample_project/Source/sample_project/Geometry/GeometryCreator.h:
```diff
@@ -0,0 +1,139 @@
+
+#pragma once
+
+#include "ArcGISMapsSDK/Components/ArcGISMapComponent.h"
+#include "CoreMinimal.h"
+#include "GameFramework/Actor.h"
+#include "Components/TextBlock.h"
+#include "ArcGISMapsSDK/Components/ArcGISLocationComponent.h"
+#include "ArcGISMapsSDK/Utils/GeoCoord/GeoPosition.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISPolygonBuilder.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISEnvelopeBuilder.h"
+#include "ArcGISMapsSDK/API/GameEngine/Geometry/ArcGISEnvelope.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISAreaUnit.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISLinearUnit.h"
+#include "Components/LineBatchComponent.h"
+#include "DrawDebugHelpers.h"
+#include "Engine/World.h"
+#include "Kismet/GameplayStatics.h"
+#include "Blueprint/UserWidget.h"
+#include "Components/ComboBoxString.h"
+#include "Components/SplineMeshComponent.h"
+#include "Engine/StaticMeshActor.h"
+#include "Materials/MaterialExpressionConstant3Vector.h"
+#include "ArcGISMapsSDK/API/GameEngine/Geometry/ArcGISGeometry.h"
+#include "ArcGISMapsSDK/Actors/ArcGISMapActor.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISAngularUnit.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISAngularUnitId.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISGeometryEngine.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISPoint.h"
+#include "ArcGISMapsSDK/API/GameEngine/Geometry/ArcGISPoint.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISSpatialReference.h"
+#include "Runtime/Engine/Classes/GameFramework/PlayerController.h"
+#include "sample_project/Routing/Breadcrumb.h"
+#include "sample_project/Routing/RouteMarker.h"
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISGeodeticDistanceResult.h"
+#include "GeometryCreator.generated.h"
```
Do you need all of these? This seems like a lot. The ones you do need please sort them.

Jade-JadeH replied:
Removed some

mmgaw113 commented on sample_project/Source/sample_project/Geometry/GeometryCreator.h:
```diff
@@ -0,0 +1,139 @@
+#include "ArcGISMapsSDK/BlueprintNodes/GameEngine/Geometry/ArcGISGeodeticDistanceResult.h"
+#include "GeometryCreator.generated.h"
+
+class UArcGISMapComponent;
+class UArcGISCameraControllerComponent;
+class UStaticMeshComponent;
+class UArcGISLocationComponent;
+class ULineBatchComponent;
```
Sort.

mmgaw113 commented on sample_project/Source/sample_project/Geometry/GeometryCreator.h:
```diff
@@ -0,0 +1,139 @@
+public:
+	AGeometryCreator();
+
+	UPROPERTY(EditAnywhere, BlueprintReadWrite)
+	TObjectPtr<UUserWidget> UIWidget;
+	TObjectPtr<UComboBoxString> UnitDropdown;
+	UPROPERTY(BlueprintReadWrite, EditAnywhere)
+	float Calculation = 0.0;
+	UFUNCTION(BlueprintCallable)
+	void SetCalculation(float InCalculation);
+	UFUNCTION(BlueprintCallable)
+	float GetCalculation();
+	UFUNCTION(BlueprintCallable)
+	void SetLinearUnit(UArcGISLinearUnit* LU);
+	UFUNCTION(BlueprintCallable)
+	void SetAreaUnit(UArcGISAreaUnit* AU);
+	UFUNCTION(BlueprintCallable)
+	UArcGISLinearUnit* GetLinearUnit();
+	UFUNCTION(BlueprintCallable)
+	UArcGISAreaUnit* GetAreaUnit();
+
```
Sort

mmgaw113 commented on sample_project/Source/sample_project/Geometry/GeometryCreator.h:
```diff
@@ -0,0 +1,139 @@
+	bool bIsDragging = false;
+	float InterpolationInterval = 10000;
+	FString UnitText;
+	FVector2D RouteCueScale = FVector2D(5);
+	UStaticMesh* RouteMesh = LoadObject<UStaticMesh>(nullptr, TEXT("/Game/SampleViewer/SharedResources/Geometries/Cube.Cube"));
+	TDoubleLinkedList<USplineMeshComponent*> SplineMeshComponents;
+	UPROPERTY(EditAnywhere, meta = (AllowPrivateAccess))
+	AArcGISMapActor* MapActor;
+	UPROPERTY(EditAnywhere, meta = (AllowPrivateAccess))
+	TObjectPtr<UArcGISMapComponent> MapComponent;
+	UPROPERTY()
+	FVector LastRootPosition;
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, meta = (AllowPrivateAccess))
+	bool bIsEnvelopeMode = false;
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, meta = (AllowPrivateAccess))
+	bool bIsPolygonMode = false;
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, meta = (AllowPrivateAccess))
+	bool bIsPolylineMode = true;
+	UPROPERTY(BlueprintReadWrite, EditAnywhere, meta = (AllowPrivateAccess))
+	class AInputManager* inputManager;
+	UPROPERTY(EditAnywhere)
+	TArray<AActor*> FeaturePoints;
+	UPROPERTY(EditAnywhere)
+	TArray<AActor*> lastToStartInterpolationPoints;
+	UPROPERTY(EditAnywhere)
+	TArray<ARouteMarker*> Stops;
+	UPROPERTY(EditAnywhere)
+	TSubclassOf<AActor> LineMarkerPrefab;
+	UPROPERTY(EditAnywhere)
+	TSubclassOf<AActor> InterpolationMarkerPrefab;
+	UPROPERTY(EditAnywhere)
+	TObjectPtr<UArcGISPoint> StartPoint;
+	UPROPERTY(EditAnywhere, BlueprintReadOnly, meta = (AllowPrivateAccess))
+	TSubclassOf<UUserWidget> UIWidgetClass;
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, meta = (AllowPrivateAccess))
+	TObjectPtr<UArcGISAreaUnit> CurrentAreaUnit;
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, meta = (AllowPrivateAccess))
+	TObjectPtr<UArcGISLinearUnit> CurrentLinearUnit;
+	UPROPERTY()
+	UArcGISSpatialReference* SpatialReference;
```
Sort.

mmgaw113 commented on sample_project/Source/sample_project/Geometry/GeometryCreator.h:
```diff
@@ -0,0 +1,139 @@
+	UPROPERTY()
+	UArcGISSpatialReference* SpatialReference;
+
+	void CreateAndCalculatePolygon();
+	void CreateAndCalculateEnvelope(UArcGISPoint* Start, UArcGISPoint* End);
+	void Interpolate(AActor* Start, AActor* End, TArray<AActor*>& Points);
+	void VisualizeEnvelope(double MinX, double MinY, double MaxX, double MaxY, UArcGISSpatialReference* SpatialRef);
+	void UpdateDraggingVisualization();
+
+	UFUNCTION()
+	void StartGeometry();
+
+	UFUNCTION()
+	void EndGeometry();
+
+	UFUNCTION(BlueprintCallable)
+	void RenderLine(TArray<AActor*>& Points);
+
+	UFUNCTION(BlueprintCallable)
+	void ClearLine();
+
+	UFUNCTION()
+	void UpdateDisplay(float Value);
```
Sort.

mmgaw113 commented on sample_project/Content/SampleViewer/cpp_SamplesDataTable.uasset:
Please alphabetize the added sample

mmgaw113 replied:
Last I checked, G comes before T in the alphabet...

mmgaw113 commented on sample_project/Source/sample_project/Geometry/GeometryCreator.h:
```diff
@@ -0,0 +1,139 @@
+// /* Copyright 2023 Esri* * Licensed under the Apache License Version 2.0 (the "License"); * you may not use this file except in compliance with the License. * You may obtain a copy of the License at * *     http://www.apache.org/licenses/LICENSE-2.0 * * Unless required by applicable law or agreed to in writing, software * distributed under the License is distributed on an "AS IS" BASIS, * WITHO ...
```
Change copy right year

mmgaw113 commented on sample_project/Source/sample_project/Geometry/GeometryCreator.cpp:
```diff
@@ -0,0 +1,459 @@
+// /* Copyright 2023 Esri* * Licensed under the Apache License Version 2.0 (the "License"); * you may not use this file except in compliance with the License. * You may obtain a copy of the License at * *     http://www.apache.org/licenses/LICENSE-2.0 * * Unless required by applicable law or agreed to in writing, software * distributed under the License is distributed on an "AS IS" BASIS, * WITHO ...
```
Change copy right year.

Jade-JadeH replied:
I changed to 2025 for this sample purpose

mmgaw113 commented on sample_project/Source/sample_project/Geometry/GeometryCreator.cpp:
alphabetize and format functions

Jade-JadeH replied:
All of the function names are uppercamelcase

mmgaw113 commented on sample_project/Content/SampleViewer/Samples/Geometry/UserInterface/UI_Geometry.uasset:
![{C01DA443-423C-43DE-811B-6E6F86265FB9}](https://github.com/user-attachments/assets/9b877eda-f88b-45ff-8869-9feb9e154b94)
Is there a way to clean this up please? It is a bit hard to read.

mmgaw113 commented on sample_project/Content/Sample_projectGameMode_BP.uasset:
I'm not sure I see the difference between main and your branch... The images your provided look identical.

mmgaw113 reviewed (changes requested):
See comments.

Jade-JadeH commented on sample_project/Source/sample_project/InputManager.cpp:
```diff
@@ -15,6 +15,15 @@ AInputManager::AInputManager()
 void AInputManager::BeginPlay()
 {
 	Super::BeginPlay();
+
+	MappingContext = LoadObject<UInputMappingContext>(
+		nullptr, TEXT("/Script/EnhancedInput.InputMappingContext'/Game/SampleViewer/SharedResources/Input/IAC_SamplesInput.IAC_SamplesInput'"));
+
+	MousePress = LoadObject<UInputAction>(
+		nullptr, TEXT("/Script/EnhancedInput.InputAction'/Game/SampleViewer/SharedResources/Input/IA_LeftMouseClick.IA_LeftMouseClick'"));
+
+	ShiftModifier = LoadObject<UInputAction>(
+		nullptr, TEXT("/Script/EnhancedInput.InputAction'/Game/SampleViewer/SharedResources/Input/IA_Shift.IA_Shift'"));
```
This is for code safety purpose. It's not recommended to initialize complex variables in .h file since it sometimes fires too early in header file before UObject system is fully initialized, and this can cause crash. In this case, it doesn't crash but it's not good code practice.

mmgaw113 replied:
Then can it be moved to the constructor instead of begin play?

Jade-JadeH replied:
it can still leads to null pointer in constructors, in begin play, everything is loaded and is 0 risk

avr757 reviewed (changes requested):
Great to see a Geometry sample for Unreal! A really good start! Nice work!
I have some UI/UX comments:

1. There is a gap between the icons and the menu - should be none
![image](https://github.com/user-attachments/assets/27b35824-08ed-4ba8-bbea-9adc1a9b0d8b)
Also, I see a vertical misalignment
![image](https://github.com/user-attachments/assets/acec862a-936d-4743-96d8-7bbac3b7d312)
2. The bottom of the menu is missing a purple accent
![image](https://github.com/user-attachments/assets/1b250d77-7ea5-4b42-9fed-2c4926cdf812)
It should have a purple line as the design
![image](https://github.com/user-attachments/assets/e00dd9a9-a4bd-4fee-91c9-4b057d183737)
3. The text "Result" and "Unit" should be 2 or more sizes smaller, bold, in a grey ish color. Also, the margin on top is much more smaller than the one in the bottom (as highlighted). Top margin and bot margin should be equal. 
![image](https://github.com/user-attachments/assets/536262a9-56ec-4c06-8999-76d41c92ba26)
4.  It seems that the each of the unit button box is not centered? 
![image](https://github.com/user-attachments/assets/e62a980b-2023-45a3-87f6-1000712c34e0)
Sometimes when it even goes out of the text
![image](https://github.com/user-attachments/assets/6deea65b-70ec-466f-b30e-19ac4f596540)
Also, is it possible to make the unit button an actual button as the design?
![image](https://github.com/user-attachments/assets/0d2efe17-1668-4cbd-8c37-933dbb4e9e81)
For UX, it should also have a button highlight when the mouse cursor is hovered over on these buttons (like the one in Unity)
![image](https://github.com/user-attachments/assets/b397fa37-8b1a-461d-a101-0bf4ea4bd5b8)
5. I noticed each of these 3 background colors are different, there only should be 2 colors actually. Menu color and hover color.
![image](https://github.com/user-attachments/assets/727bf880-091e-4692-8f96-b5acdce4d485)
Please check the design for reference in terms of background color consistency. 
![image](https://github.com/user-attachments/assets/a826085d-fe24-4594-968d-05c24c2d3fec)
6. The Reset button should be moved down to be align with the Unit buttons vertically. The info icon can be moved up slightly.
![image](https://github.com/user-attachments/assets/c0f3bf21-425c-4ddf-9dad-f6535005fafb)

Jade-JadeH commented:
not seeing a gap... @avr757 

![Capture](https://github.com/user-attachments/assets/86189edb-1fe3-43d0-a706-52440fd6e8c9)

avr757 reviewed (changes requested):
Wow thanks for making all the changes and fixing bugs. It looks much better now and there are lots of improvements!

I still noticed a few quick things:
1. Could you increase the result text size to 16, and make the font bold?
![image](https://github.com/user-attachments/assets/87115035-4e9e-4227-9a39-0a2015efb8f9)
![image](https://github.com/user-attachments/assets/1e06ac27-d6bf-4f5a-b1b8-a0f4962971ec)
2. I think the reset button will look better if it moves up vertically a bit. I think `Pos y = 160` is a sweet spot.
![image](https://github.com/user-attachments/assets/5a6105aa-bfb0-4f9f-bbd1-c2ccecd76824)
![image](https://github.com/user-attachments/assets/c2d17092-fcfb-4641-a45b-dd551a08012b)
3. This is not a huge deal but it seems the gap still exists. Could you check and test it in full screen mode? I think this happens in full screen mode actually. Not sure how easy is it to fix it. But don't sweat it.
![image](https://github.com/user-attachments/assets/425e5e14-7e5a-402b-bb2a-1ca820018914)
5. I feel like there is still a mismatch between the geometry buttons and the menu. Is there a way to use color picker to match those colors? Just wondering.
![image](https://github.com/user-attachments/assets/17afc684-6283-45a5-8697-593cd0ad834d)
6. Often times I noticed that, even though envelope is selected as the geometry, when selecting a region, it shows a triangle instead. After resetting, I noticed a bug with:
`UnrealEditor_sample_project!AGeometryCreator::ClearLine() [C:\Users\Documents\GitHub\arcgis-maps-sdk-unreal-engine-samples\sample_project\Source\sample_project\Geometry\GeometryCreator.cpp:253]`
![image](https://github.com/user-attachments/assets/04d9708f-de43-40fb-aced-943c7312f00b)

Jade-JadeH commented:
@avr757 Fixed the UI issues you mentioned. Fixed the gap issue which appears to be an anchor issue. Fixed the triangle issue (at least on my computer...) by adding a vertices check.

avr757 reviewed (approved):
UI looks great now! Thanks for all your efforts addressing the feedback! :D

![image](https://github.com/user-attachments/assets/c44533bb-75c9-4aae-aa33-112940faa04d)
